src/main/java/org/folio/circulation/infrastructure/storage/CheckOutLockRepository.java

```diff
@@ -0,0 +1,84 @@
+package org.folio.circulation.infrastructure.storage;
+
+import io.vertx.core.Vertx;
+import io.vertx.core.json.JsonObject;
+import io.vertx.ext.web.RoutingContext;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.folio.circulation.domain.CheckOutLock;
+import org.folio.circulation.domain.LoanAndRelatedRecords;
+import org.folio.circulation.support.Clients;
+import org.folio.circulation.support.CollectionResourceClient;
+import org.folio.circulation.support.http.client.Response;
+import org.folio.circulation.support.http.client.ResponseInterpreter;
+import org.folio.circulation.support.results.Result;
+
+import java.lang.invoke.MethodHandles;
+import java.util.concurrent.CompletableFuture;
+
+import static org.folio.circulation.support.http.ResponseMapping.forwardOnFailure;
+import static org.folio.circulation.support.http.ResponseMapping.mapUsingJson;
+
+public class CheckOutLockRepository {
+  private static final Logger log = LogManager.getLogger(MethodHandles.lookup().lookupClass());
+  private final CollectionResourceClient checkOutLockClient;
+  private final Vertx vertx;
+
+  public CheckOutLockRepository(Clients clients, RoutingContext routingContext) {
+    this.checkOutLockClient = clients.checkOutLockClient();
+    this.vertx = routingContext.vertx();
+  }
+
+  public void createLockWithRetry(int noOfAttempts, CompletableFuture<CheckOutLock> future, LoanAndRelatedRecords records) {
+    log.debug("createLockWithRetry:: Retrying lock creation {} ", noOfAttempts);
+    int maxRetryAttempts = records.getCheckoutLockConfiguration().getNoOfRetryAttempts();
+    int retryInterval = records.getCheckoutLockConfiguration().getRetryInterval();
+    try {
+      create(records)
+        .whenComplete((res, err) -> {
+          if (res.succeeded()) {
+            log.info("createLockWithRetry:: completing future successfully");
+            future.complete(res.value());
+          } else {
+            if (noOfAttempts <= maxRetryAttempts) {
+              log.info("createLockWithRetry:: Retry attempt {} for lock creation with delay {}", noOfAttempts, retryInterval);
+              vertx.setTimer(retryInterval, h -> createLockWithRetry(noOfAttempts + 1, future, records));
+            } else {
+              String error = res.cause() != null ? res.cause().toString() : "";
+              log.warn("createLockWithRetry:: Completing exceptionally {} ", error);
+              future.completeExceptionally(new RuntimeException(error));
```

Comment:
It might be better to define a `CheckoutLockException`. Is there any information at this point that could be added to the exception. Like the lock id perhaps? If not, I still think having our own exception type at least sends a message that we have thought through the ways this can fail, even if they are edge cases.

Think of it like this: imagine trying to debug a prod system. If the error logged isn't a known `CheckoutLockException` that's meaningful.

Reply:
Since we aren't doing anything based on the exception type. I feel that runtime exception wont create any harm. I logged the itemId to which we aren't able to acquire the lock in the caller method. The caller exception will looks like something below.
**ERROR OutByBarcodeResource acquireLock:: Unable to acquire lock for item item7
Tue, Jun 27 2023 6:07:01 pm | java.lang.RuntimeException: org.folio.circulation.support.ForwardOnFailure@4d4ab6e[failureResponse=Response from "http://okapi:9130/check-out-lock-storage" status code: 503 body: "Unable to acquire lock", content type: "text/plain"].**
 Please let me know if you still feel we need to have a custom exception.